{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "import os,re,subprocess,copy\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from natsort import natsorted\n",
    "\n",
    "path = '/data/s3cha/CHO_ENOSI_JOB/e22076b9e7ab40ef8fc2c63eddf67eba'\n",
    "event = path + '/all_events.txt'\n",
    "ref_fa = '/data/genome/hamster/picr_old/picr.fa'\n",
    "\n",
    "pasa_path = '/data/shangzhong/Picr_assembly/Annotation/PASA/pasa_stringtie'\n",
    "pasa_assm_fa = pasa_path + '/pasa_assem/picr_db_stringtie.assemblies.fasta'\n",
    "pasa_gff = pasa_path + '/pasa_assem/picr_db_stringtie.pasa_assemblies.gff3'\n",
    "gff = pasa_path + '/03_pasa_stringtie.gff3'\n",
    "gff_pr = pasa_path + '/03_pasa_stringtie_pr.fa'\n",
    "\n",
    "transde_gff = pasa_path + '/transdecoder/picr_db_stringtie.transdecoder.genome.gff3'\n",
    "transde_pep = pasa_path + '/transdecoder/picr_db_stringtie.assemblies.fasta.transdecoder.pep'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  For transcripts whose confidence level equal to 4 or 5, we check if there is evidence (proteomics or riboseq) to support them, if not, remove their CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "uniq_event = path + '/all_events_uniq.txt'\n",
    "if not os.path.exists(uniq_event):\n",
    "    df = pd.read_csv(event,sep='\\t',header=0)\n",
    "    df = df[['Event','Peptide','Chromosome','Location','Strand']].drop_duplicates()\n",
    "    df.to_csv(uniq_event,sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# get gene assemble\n",
    "all_gff = {} # {geneid:[gene_line],{asm_id:[[asm_id_line]]}}\n",
    "gene_rna_dic = {}  # {geneid:[[gene_line],[asm_ids]]}\n",
    "rna_all_dic = {} # {asmbl_id:[[line1],['line2]]}\n",
    "gene_pos = {}  # {geneid:[chr,strand,start,end]}\n",
    "with open(gff) as in_f:\n",
    "    for line in in_f:\n",
    "        if line.startswith('#'):continue\n",
    "        item = line.strip().split('\\t')\n",
    "        if item[2] == 'gene':\n",
    "            geneid = re.search('(?<=gene_id=).+?(?=;)',item[8]).group(0)\n",
    "            gene_pos[geneid] = [item[0],item[6],int(item[3]),int(item[4])]\n",
    "            gene_rna_dic[geneid]=[item,[]]\n",
    "            all_gff[geneid] = [item,{}]\n",
    "        elif item[2] in ['mRNA','lncRNA']:\n",
    "            rnaid = re.search('(?<=ID=).+?(?=;)',item[8]).group(0)\n",
    "            rna_all_dic[rnaid] = [item]\n",
    "            chrom = item[0]\n",
    "            strand = item[6]\n",
    "            geneid = re.search('(?<=gene_id=).+?(?=;)',item[8]).group(0)\n",
    "            gename = re.search('(?<=Name=).+?(?=;|$)',item[8]).group(0)\n",
    "            gene_rna_dic[geneid][1].append(rnaid)\n",
    "            all_gff[geneid][1][rnaid] = [item]\n",
    "        else:\n",
    "            rna_all_dic[rnaid].append(item)\n",
    "            all_gff[geneid][1][rnaid].append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# # get riboseq expression to filter the reads\n",
    "# ribo = '/data/shangzhong/RibosomeProfiling/previous_data'\n",
    "# folders = natsorted([f for f in os.listdir(ribo) if f.endswith('quant')])\n",
    "\n",
    "# assm_ribo_expressed  = []\n",
    "# for f in folders:\n",
    "#     df = pd.read_csv(ribo+'/'+f+'/quant.sf',sep='\\t',header=0)\n",
    "#     assms = df.query('TPM>=1')['Name'].tolist()\n",
    "#     assm_ribo_expressed.extend(assms)\n",
    "# assm_ribo_expressed = list(set(assm_ribo_expressed))\n",
    "\n",
    "# get proteogenomics support\n",
    "known_pep = '/data/s3cha/CHO_ENOSI_JOB/e22076b9e7ab40ef8fc2c63eddf67eba/known_asmbl.txt'\n",
    "pep_assm = pd.read_csv(known_pep,sep='\\t',header=None)[0].tolist()\n",
    "proved_assm = list(set(pep_assm)) # +assm_ribo_expressed\n",
    "\n",
    "low_qal_rna = []\n",
    "rna_gid_dic = {}\n",
    "with open(gff) as f:\n",
    "    for line in f:\n",
    "        if line.startswith('#'):continue\n",
    "        item = line.strip().split('\\t')\n",
    "        if item[2] == 'mRNA':\n",
    "            cs = item[8][-1]\n",
    "            gid = re.search('(?<=gene_id=).+?(?=;)',item[8]).group(0)\n",
    "            assm = re.search('(?<=ID=).+?(?=;)',item[8]).group(0)\n",
    "            rna_gid_dic[assm] = gid\n",
    "            if cs in ['4','5']:\n",
    "                low_qal_rna.append(assm)\n",
    "\n",
    "fail_assm = set(low_qal_rna)-set(proved_assm)#-set(assm_ribo_expressed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Therea are 12325 transcripts whose confidence scores are 4 or 5 and don't have spectrum to support them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# remove the low quality proteins\n",
    "for gene,v in all_gff.iteritems():\n",
    "    for assm in v[1]:\n",
    "        if assm in fail_assm:\n",
    "            all_gff[gene][1][assm] = [item for item in all_gff[gene][1][assm] if item[2]!='CDS']\n",
    "\n",
    "# output the filtered annotation\n",
    "fil_gff = pasa_path+'/04_fil_pasa_stringtie.gff3'\n",
    "if not os.path.exists(fil_gff):\n",
    "    with open(gff) as in_f,open(fil_gff,'w') as out_f:\n",
    "        for line in in_f:\n",
    "            if line.startswith('#'):\n",
    "                out_f.write(line)\n",
    "                continue\n",
    "            item = line.strip().split('\\t')\n",
    "            if item[2] != 'CDS':\n",
    "                out_f.write(line)\n",
    "            else:\n",
    "                assm = re.search('(?<=Parent=).+?(?=;)',line).group(0)\n",
    "                if assm in fail_assm:\n",
    "                    continue\n",
    "                else:\n",
    "                    out_f.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# transdecoder protein sequence\n",
    "trans_pr_pos_dic = {}\n",
    "with open(transde_gff) as f:\n",
    "    for line in f:\n",
    "        if line.strip() == '': continue\n",
    "        item = line.split('\\t')\n",
    "        chrom = item[0]\n",
    "        start = int(item[3])\n",
    "        end = int(item[4])\n",
    "        strand = item[6]\n",
    "        anno = item[8]\n",
    "        if item[2] =='CDS':\n",
    "            fid = re.search('(?<=ID=).+?(?=;)',anno).group(0)[4:]\n",
    "            if fid not in trans_pr_pos_dic:\n",
    "                trans_pr_pos_dic[fid] = [chrom+strand,[start,end]]\n",
    "            else:\n",
    "                trans_pr_pos_dic[fid].append([start,end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# read pasa rna sequence\n",
    "pasa_assm_dic = SeqIO.index(pasa_assm_fa,'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# read pasa assembled rna position\n",
    "pasa_pos_dic = {}  # {assm:[[start,end]]}\n",
    "with open(pasa_gff) as f:\n",
    "    for line in f:\n",
    "        if line.strip() == '' or line.startswith('#'):\n",
    "            continue\n",
    "        item = line.strip().split('\\t')\n",
    "        chrom = item[0]\n",
    "        start = int(item[3])\n",
    "        end = int(item[4])\n",
    "        strand = item[6]\n",
    "        anno = item[8]\n",
    "        tr = re.search('(?<=Target=).+?(?=\\ )',anno).group(0)\n",
    "        if tr not in pasa_pos_dic:\n",
    "            pasa_pos_dic[tr] = [chrom+strand,[start,end]]\n",
    "        else:\n",
    "            pasa_pos_dic[tr].append([start,end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# read gff protein sequence\n",
    "gff_pr_dic = SeqIO.index(gff_pr,'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "# get protein and transcript id mapping\n",
    "transde_pep_dic = SeqIO.index(transde_pep,'fasta')\n",
    "tr_pr_dic = {}\n",
    "for prid in transde_pep_dic:\n",
    "    tr = prid.split('.')[0]\n",
    "    if tr in tr_pr_dic:\n",
    "        tr_pr_dic[tr].append(prid)\n",
    "    else:\n",
    "        tr_pr_dic[tr] = [prid]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "def genes_overlap_with_pep(item,gene_pos):\n",
    "    '''this function gets overlapped gene ids for a event line\n",
    "    * gene_pos: {geneid:[chrom,start,end]}\n",
    "    * item: splited event line\n",
    "    '''\n",
    "    pep=''.join(item[1].split(':'))\n",
    "    chrom = item[2]\n",
    "    pep_str = item[-1]\n",
    "    pos = []\n",
    "    for ps in item[3].split(';'):\n",
    "        pos.extend([int(p) for p in ps.split('-')])\n",
    "    p_s = min(pos)\n",
    "    p_e = max(pos)\n",
    "    gene_ids = []\n",
    "    for k,v in gene_pos.iteritems():\n",
    "        if chrom == v[0]:\n",
    "            if pep_str == v[1]:\n",
    "                if v[2]<p_s<p_e<v[3]:\n",
    "                    gene_ids.append(k)\n",
    "    return gene_ids,p_s,p_e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "hideCode": false,
    "hidePrompt": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def getORF(in_fa,pep):\n",
    "    '''this function predicts the ORF in rna sequence and returen the longest ORF that match\n",
    "    provided peptide'''\n",
    "    cmd = ('~/Installation/ORFfinder -in {inf} -strand plus -out inter.fa').format(inf=in_fa)\n",
    "    subprocess.call(cmd,shell=True)\n",
    "    with open('out.fa','w') as out:\n",
    "        for record in SeqIO.parse('inter.fa','fasta'):\n",
    "            if pep in record.seq:\n",
    "                SeqIO.write(record,out,'fasta')\n",
    "    os.remove('inter.fa')\n",
    "    l = 0\n",
    "    res = ''\n",
    "    for record in SeqIO.parse('out.fa','fasta'):\n",
    "        if len(record.seq) > l:\n",
    "            res = record\n",
    "            l = len(record.seq)\n",
    "    os.remove('out.fa')\n",
    "    os.remove(in_fa)\n",
    "    return res\n",
    "\n",
    "def getCDS_gff(assm_gff,pr_seq):\n",
    "    '''get cds position in gff format given protein sequence and rna gff format'''\n",
    "    strand = assm_gff[0][-1]\n",
    "    if strand == '+':\n",
    "        assm_gff[1:] = sorted(assm_gff[1:],key=lambda x:x[0])\n",
    "    else:\n",
    "        assm_gff[1:] = sorted(assm_gff[1:],key=lambda x:x[0],reverse=True)\n",
    "        assm_gff[1:] = [[p[1],p[0]] for p in assm_gff[1:]]\n",
    "    cds_gff = [assm_gff[0]]\n",
    "    head = pr_seq.id\n",
    "    item = head.split(':')\n",
    "    cds_s = int(item[1])+1\n",
    "    cds_e = int(item[2])+1\n",
    "    exon_index = [[0,0]]\n",
    "    exon_lens = [abs(l[1]-l[0])+1 for l in assm_gff[1:]]\n",
    "    for l in exon_lens:\n",
    "        pair = [exon_index[-1][-1]+1,exon_index[-1][-1]+l]\n",
    "        exon_index.append(pair)\n",
    "    exon_index = exon_index[1:]\n",
    "    for index,pos in zip(exon_index,assm_gff[1:]):\n",
    "        if index[0]<=cds_s<=index[1]:\n",
    "            if strand=='+':\n",
    "                s = pos[0] + (cds_s-index[0])\n",
    "            else:\n",
    "                s = pos[0] - (cds_s-index[0])\n",
    "        elif cds_s > index[1]:\n",
    "            continue\n",
    "        else:\n",
    "            s = pos[0]\n",
    "        if index[0]<=cds_e<=index[1]:\n",
    "            if strand=='+':\n",
    "                e = pos[0] + (cds_e-index[0])\n",
    "            else:\n",
    "                e = pos[0] - (cds_e-index[0])\n",
    "        elif cds_e < index[0]:\n",
    "            continue\n",
    "        else:\n",
    "            e = pos[1]\n",
    "        cds_gff.append([s,e])\n",
    "    if strand == '+':\n",
    "        cds_gff[1:] = sorted(cds_gff[1:],key=lambda x:x[0])\n",
    "    else:\n",
    "        cds_gff[1:] = sorted(cds_gff[1:],key=lambda x:x[0],reverse=True)\n",
    "        cds_gff[1:] = [[p[1],p[0]] for p in cds_gff[1:]]\n",
    "    return cds_gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Need to process the grouped peptides.\n",
    " This has two cases. \n",
    " 1. the new peptide maps to a transcript that already has protein, but it maps to the non coding part. the origial gene id and name remain the same. The new add gene is in the format gene_asmbl_idx.\n",
    " 2. The new peptide maps to a transcript that doesn't have protein. This is easy, we get 6 frame ORF and then extract the one that include the new peptide. Gene id and gene name doesn't change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def new_transcript_CDS(new_event_gff_dic,item,rm_genes,pasa_assm_dic,pasa_pos_dic,gene_rna_dic,rna_all_dic,gene_pos,trans_pr_pos_dic,gff_pr_dic,tr_pr_dic):\n",
    "    '''\n",
    "    * new_event_gff_dic = {} # {geneid:[[geneline],{assm:[lines]}]}\n",
    "    * item: one line of event results\n",
    "    * gene_rna_dic = {}  # {geneid:[[gene_line],[asm_ids]]}\n",
    "    * rna_all_dic = {} # {asmbl_id:[[line1],[line2]]}\n",
    "    * gene_pos = {}  # {geneid:[chr,start,end]}\n",
    "    * trans_pr_pos_dic {transdecoder_pr:[chrstr,[pos]]}\n",
    "    '''\n",
    "    pep=''.join(item[1].split(':'))\n",
    "    pep_str = item[-1]\n",
    "    update_assm = []\n",
    "    # 1. get all gene ids that overlap with the peptide\n",
    "    gene_ids,p_s,p_e = genes_overlap_with_pep(item,gene_pos)\n",
    "    # 2. for each gene id get the transcripts that has the peptide sequence\n",
    "    match_prs = []\n",
    "    for gene in gene_ids:\n",
    "        rm_genes.append(gene)\n",
    "        new_event_gff_dic[gene]=[gene_rna_dic[gene][0],{}]\n",
    "        assm_ids = gene_rna_dic[gene][1]\n",
    "        gene_cds_pos = []\n",
    "        cds_poses = []\n",
    "        cs = []\n",
    "        for assm in assm_ids:\n",
    "            '''if assm already has CDS, then update gene position to min and max of all CDS pos\n",
    "            gene name does not change.'''\n",
    "            assm_cds_pos = []\n",
    "            cds_poses.append([])\n",
    "            # get boundary for all CDS\n",
    "            known_assm_pos = rna_all_dic[assm]\n",
    "            cs.append(int(known_assm_pos[0][-1][-1]))\n",
    "            if assm in tr_pr_dic:  # assemblies have coding protein\n",
    "                prs = tr_pr_dic[assm] # dic from transdecoder pep\n",
    "                for pr in prs:\n",
    "                    pr_tr = pr.split('.')[0]\n",
    "                    if pr_tr in gff_pr_dic:  # the protein is in the annotation\n",
    "                        if pep in gff_pr_dic[pr_tr].seq:\n",
    "                            print pep,'peptide already in known sequence',pr\n",
    "                            # this is because the new peptide is the beginning of a database protein,\n",
    "                            # but not the real beginning of known protein. It was not identified in known database.\n",
    "                            break\n",
    "                    if pep in transde_pep_dic[pr].seq:\n",
    "                        match_prs.append(pr)\n",
    "                    # get cds start and end\n",
    "                    for line in known_assm_pos:\n",
    "                        if line[2] == 'CDS':\n",
    "                            gene_cds_pos.extend([int(line[3]),int(line[4])])\n",
    "                            cds_poses[-1].append((int(line[3]),int(line[4])))\n",
    "                    if cds_poses[0] != []: # has proteins, then split out the known protein as new genes\n",
    "                        new_event_gff_dic[gene][0][3] = min(gene_cds_pos)\n",
    "                        new_event_gff_dic[gene][0][4] = max(gene_cds_pos)\n",
    "        if cds_poses[0] != []: \n",
    "            '''the gene already has protines, need to split between old and new\n",
    "            change transcript id from asmbl_num to geneid_asmbl_index.\n",
    "            geneid does not change'''\n",
    "            chrom = new_event_gff_dic[gene][0][0]\n",
    "            strand = new_event_gff_dic[gene][0][6]\n",
    "            name = re.search('(?<=Name=).+?(?=;|$)',new_event_gff_dic[gene][0][8]).group(0)\n",
    "            uni_cdses = set([tuple(x) for x in cds_poses if x!=[]])\n",
    "            n = 0\n",
    "            for uni in uni_cdses:\n",
    "                n += 1\n",
    "                rnaid = gene+'_assm_'+str(n)\n",
    "                anno = 'ID={r};Parent={g};gene_id={g};Name={na};transcript_id={r};cs={cs}'.format(\n",
    "                            r=rnaid,g=gene,na=name,cs=str(max(cs)))\n",
    "                s = min(uni,key=lambda x:x[0])[0]\n",
    "                e = max(uni,key=lambda x:x[1])[1]\n",
    "                new_event_gff_dic[gene][1][rnaid] = [[chrom,'merge','mRNA',str(s),str(e),'.',strand,\n",
    "                                                         '.',anno]]\n",
    "                i = 0\n",
    "                for u in uni:\n",
    "                    i += 1\n",
    "                    anno = 'ID={r}.exon{i};Parent={r};gene_id={g};Name={na};transcript_id={r}'.format(\n",
    "                            r=rnaid,i=i,g=gene,na=name)\n",
    "                    new_event_gff_dic[gene][1][rnaid].append([chrom,'merge','exon',str(u[0]),str(u[1]),'.',\n",
    "                                strand,'.',anno])\n",
    "                i = 0\n",
    "                for u in uni:\n",
    "                    i += 1\n",
    "                    anno = 'ID={r}.CDS{i};Parent={r};gene_id={g};Name={na};protein_id={r}'.format(\n",
    "                            r=rnaid,i=i,g=gene,na=name)\n",
    "                    new_event_gff_dic[gene][1][rnaid].append([chrom,'merge','CDS',str(u[0]),str(u[1]),'.',\n",
    "                                strand,'.',anno])\n",
    "        else: # the gene doesn't have proteins\n",
    "            for assm in assm_ids:\n",
    "                known_assm_pos = [line for line in rna_all_dic[assm] if line[2]!='CDS']\n",
    "                new_event_gff_dic[gene][1][assm]=known_assm_pos\n",
    "        #------------- add new peptide to the gff file \n",
    "        if match_prs == []:# mapped gene doesn't have CDS,or have CDS but don't include pep\n",
    "            n = 1\n",
    "            for assm in assm_ids:\n",
    "                assm_record = pasa_assm_dic[assm]\n",
    "                with open(assm+'.fa','w') as f:\n",
    "                    SeqIO.write(assm_record,f,'fasta')\n",
    "                orf = getORF(assm+'.fa',pep)\n",
    "                assm_pos = copy.copy(pasa_pos_dic[assm])\n",
    "                chrom = assm_pos[0][:-1]\n",
    "                strand = assm_pos[0][-1]\n",
    "                rnaid = assm\n",
    "                gene_id = gene\n",
    "                '''here assign to gene, in the followings if need to split the original genes,\n",
    "                the gene id would be gene_asmbl_num. Otherwise, do not change gene id, directly\n",
    "                add cds '''\n",
    "                if isinstance(orf,str):  # in another transcripts\n",
    "                    continue\n",
    "                cds_pos = getCDS_gff(assm_pos,orf)\n",
    "                # write gene\n",
    "                if gene_cds_pos!=[]:\n",
    "                    print gene_id,assm,'new pep maps to it which already has coding protein'\n",
    "                    update_assm.append(assm)\n",
    "                    gene_id = 'gene_'+rnaid\n",
    "                    rnaid = rnaid+'_'+str(n)\n",
    "                    start = min(cds_pos[1:],key=lambda x:x[0])[0]\n",
    "                    end = max(cds_pos[1:],key=lambda x:x[1])[1]\n",
    "                    anno='ID={g};gene_id={g};Name={g}'.format(g=gene_id)\n",
    "                    new_event_gff_dic[gene_id] = [[chrom,'merge','gene',str(start),str(end),\n",
    "                                                '.',strand,'.',anno],{}]\n",
    "                    anno=('ID={r};Parent={g};gene_id={g};Name={g};transcript_id={r};cs=4').format(\n",
    "                                                    r=rnaid,g=gene_id)\n",
    "                    new_event_gff_dic[gene_id][1][rnaid]=[[chrom,'merge','mRNA',str(start),str(end),\n",
    "                                    '.',strand,'.',anno]]\n",
    "                    # write assm\n",
    "                    i = 0\n",
    "                    for exon_p in cds_pos[1:]:\n",
    "                        i +=1\n",
    "                        anno=('ID={r}.exon{i};Parent={r};gene_id={g};Name={g};transcript_id={r}').format(\n",
    "                                               r=rnaid,i=str(i),g=gene_id)\n",
    "                        new_event_gff_dic[gene_id][1][rnaid].append([chrom,'merge','exon',str(exon_p[0]),\n",
    "                                            str(exon_p[1]),'.',strand,'.',anno])\n",
    "                else:\n",
    "                    print gene_id,assm,'new pep maps to it without coding protein'\n",
    "                    update_assm.append(assm)\n",
    "                i = 0\n",
    "                for cds_p in cds_pos[1:]:\n",
    "                    i +=1\n",
    "                    anno=('ID={r}.cds{i};Parent={r};gene_id={g};Name={g};protein_id={r}').format(\n",
    "                                           r=rnaid,i=str(i),g=gene_id)\n",
    "                    new_event_gff_dic[gene_id][1][rnaid].append([chrom,'merge','CDS',str(cds_p[0]),\n",
    "                                        str(cds_p[1]),'.',strand,'.',anno])\n",
    "            n += 1\n",
    "        elif len(match_prs) > 1:\n",
    "            print pep + ' maps to two or more rnas'\n",
    "        else:\n",
    "            for match_pr in match_prs[0:1]: # predicted protein has the peptide\n",
    "                n = 1\n",
    "                assm = match_pr.split('.')[0]\n",
    "                print assm\n",
    "                if gene_cds_pos != []: # known gff already has protein\n",
    "                    known_cds_s,known_cds_e = min(gene_cds_pos),max(gene_cds_pos)\n",
    "                    if p_s<=known_cds_s<=p_e or known_cds_s<=p_s<=known_cds_e:\n",
    "                        print match_pr + ' new peptide overlap with known'\n",
    "                        continue\n",
    "                    transde_pr_pos = trans_pr_pos_dic[match_pr]\n",
    "                    # create new gff\n",
    "                    chrom = transde_pr_pos[0][:-1]\n",
    "                    strand = transde_pr_pos[0][-1]\n",
    "                    rnaid = assm+'_'+str(n)\n",
    "                    gene_id = 'gene_'+rnaid\n",
    "                    all_pos = [p for cds_p in transde_pr_pos[1:] for p in cds_p]\n",
    "                    start = min(all_pos);end=max(all_pos)\n",
    "                    anno='ID={g};gene_id={g};Name={g}'.format(g=gene_id)\n",
    "                    new_event_gff_dic[gene_id] = [[chrom,'merge','gene',str(start),str(end),\n",
    "                                                '.',strand,'.',anno],{}]\n",
    "                    anno=('ID={r};Parent={g};gene_id={g};Name={g};transcript_id={r}').format(\n",
    "                                                    r=rnaid,g=gene_id)\n",
    "                    new_event_gff_dic[gene_id][1][rnaid]=[[chrom,'merge','mRNA',str(start),str(end),\n",
    "                                    '.',strand,'.',anno]]\n",
    "                    i = 0\n",
    "                    for cds_p in transde_pr_pos[1:]:\n",
    "                        i += 1\n",
    "                        anno=('ID={r}.exon{i};Parent={r};gene_id={g};Name={g};transcript_id={r}').format(\n",
    "                                           r=rnaid,i=str(i),g=gene_id)\n",
    "                        new_event_gff_dic[gene_id][1][rnaid].append([chrom,'merge','exon',str(cds_p[0]),\n",
    "                                            str(cds_p[1]),'.',strand,'.',anno])\n",
    "                        anno=('ID={r}.cds{i};Parent={r};gene_id={g};Name={g};protein_id={r}').format(\n",
    "                                           r=rnaid,i=str(i),g=gene_id)\n",
    "                        new_event_gff_dic[gene_id][1][rnaid].append([chrom,'merge','CDS',str(cds_p[0]),\n",
    "                                            str(cds_p[1]),'.',strand,'.',anno])\n",
    "            n += 1\n",
    "    return new_event_gff_dic,rm_genes,update_assm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_1415 asmbl_4507 new pep maps to it without coding protein\n",
      "gene_11811_1 asmbl_33799 new pep maps to it without coding protein\n",
      "gene_28383 asmbl_80128 new pep maps to it without coding protein\n",
      "gene_4459 asmbl_13136 new pep maps to it which already has coding protein\n",
      "gene_4459 asmbl_13138 new pep maps to it which already has coding protein\n",
      "gene_4459 asmbl_13139 new pep maps to it which already has coding protein\n",
      "gene_4666 asmbl_13701 new pep maps to it without coding protein\n",
      "gene_4666 asmbl_13702 new pep maps to it without coding protein\n",
      "gene_16694 asmbl_47213 new pep maps to it without coding protein\n",
      "gene_28567 asmbl_80639 new pep maps to it without coding protein\n",
      "gene_29483 asmbl_83539 new pep maps to it which already has coding protein\n",
      "gene_29483 asmbl_83541 new pep maps to it which already has coding protein\n",
      "gene_17008_1 asmbl_48309 new pep maps to it without coding protein\n",
      "gene_17008_1 asmbl_48287 new pep maps to it without coding protein\n",
      "gene_17008_4 asmbl_48305 new pep maps to it which already has coding protein\n",
      "gene_17008_3 asmbl_48289 new pep maps to it which already has coding protein\n",
      "gene_17008_2 asmbl_48286 new pep maps to it without coding protein\n",
      "gene_17008_2 asmbl_48288 new pep maps to it without coding protein\n",
      "gene_17008_2 asmbl_48284 new pep maps to it without coding protein\n",
      "gene_29278 asmbl_83034 new pep maps to it without coding protein\n",
      "gene_6616_1 asmbl_19762 new pep maps to it without coding protein\n",
      "gene_6616_3 asmbl_19767 new pep maps to it without coding protein\n",
      "gene_6616_2 asmbl_19763 new pep maps to it without coding protein\n",
      "gene_13153_1 asmbl_37912 new pep maps to it which already has coding protein\n",
      "HVGMAVAGLLADAR peptide already in known sequence asmbl_39660.p1\n",
      "HVGMAVAGLLADAR peptide already in known sequence asmbl_39654.p1\n",
      "gene_13839 asmbl_39660 new pep maps to it which already has coding protein\n",
      "gene_13839 asmbl_39654 new pep maps to it which already has coding protein\n",
      "gene_13839 asmbl_39659 new pep maps to it which already has coding protein\n",
      "gene_10032 asmbl_29272 new pep maps to it without coding protein\n",
      "gene_26940 asmbl_76148 new pep maps to it without coding protein\n",
      "gene_622 asmbl_2052 new pep maps to it without coding protein\n",
      "gene_26570_1 asmbl_75057 new pep maps to it without coding protein\n",
      "gene_25083 asmbl_70902 new pep maps to it without coding protein\n",
      "gene_30486 asmbl_86301 new pep maps to it without coding protein\n",
      "gene_30486 asmbl_86302 new pep maps to it without coding protein\n",
      "gene_11258_1 asmbl_32414 new pep maps to it without coding protein\n",
      "gene_11258_1 asmbl_32415 new pep maps to it without coding protein\n",
      "gene_11258_1 asmbl_32417 new pep maps to it without coding protein\n",
      "gene_21243 asmbl_59673 new pep maps to it without coding protein\n",
      "gene_14471 asmbl_41314 new pep maps to it without coding protein\n",
      "gene_5572 asmbl_16135 new pep maps to it without coding protein\n",
      "gene_5572 asmbl_16136 new pep maps to it without coding protein\n",
      "gene_28757 asmbl_81468 new pep maps to it which already has coding protein\n",
      "gene_28757 asmbl_81469 new pep maps to it which already has coding protein\n",
      "gene_28974 asmbl_82117 new pep maps to it without coding protein\n",
      "gene_17222 asmbl_48767 new pep maps to it without coding protein\n",
      "gene_17222 asmbl_48768 new pep maps to it without coding protein\n",
      "gene_6546 asmbl_19498 new pep maps to it without coding protein\n",
      "gene_11884 asmbl_33992 new pep maps to it without coding protein\n",
      "gene_11884 asmbl_33993 new pep maps to it without coding protein\n",
      "gene_6188 asmbl_18163 new pep maps to it without coding protein\n",
      "gene_13841 asmbl_39657 new pep maps to it without coding protein\n",
      "gene_13841 asmbl_39656 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68585 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68586 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68587 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68588 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68589 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68590 new pep maps to it without coding protein\n",
      "gene_1982 asmbl_6000 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61079 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61080 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61081 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21010 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21009 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21011 new pep maps to it without coding protein\n",
      "gene_20729 asmbl_58324 new pep maps to it without coding protein\n",
      "gene_7510 asmbl_22087 new pep maps to it without coding protein\n",
      "gene_3507_2 asmbl_10317 new pep maps to it without coding protein\n",
      "gene_3507_2 asmbl_10318 new pep maps to it without coding protein\n",
      "gene_16694 asmbl_47213 new pep maps to it without coding protein\n",
      "gene_23151 asmbl_65485 new pep maps to it without coding protein\n",
      "gene_23151 asmbl_65486 new pep maps to it without coding protein\n",
      "gene_23151 asmbl_65488 new pep maps to it without coding protein\n",
      "gene_23151 asmbl_65487 new pep maps to it without coding protein\n",
      "gene_23151 asmbl_65489 new pep maps to it without coding protein\n",
      "gene_7356 asmbl_21720 new pep maps to it without coding protein\n",
      "gene_27564 asmbl_77738 new pep maps to it without coding protein\n",
      "gene_28567 asmbl_80639 new pep maps to it without coding protein\n",
      "gene_25621 asmbl_72228 new pep maps to it without coding protein\n",
      "gene_11615 asmbl_33284 new pep maps to it without coding protein\n",
      "gene_17008_1 asmbl_48309 new pep maps to it without coding protein\n",
      "gene_17008_1 asmbl_48287 new pep maps to it without coding protein\n",
      "gene_17008_4 asmbl_48305 new pep maps to it which already has coding protein\n",
      "gene_17008_3 asmbl_48289 new pep maps to it which already has coding protein\n",
      "gene_17008_2 asmbl_48286 new pep maps to it without coding protein\n",
      "gene_17008_2 asmbl_48288 new pep maps to it without coding protein\n",
      "gene_17008_2 asmbl_48284 new pep maps to it without coding protein\n",
      "gene_29747 asmbl_84337 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52487 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52488 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52486 new pep maps to it without coding protein\n",
      "gene_21976 asmbl_62067 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66255 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66256 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66259 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66260 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66261 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66262 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66263 new pep maps to it without coding protein\n",
      "gene_23002 asmbl_65039 new pep maps to it without coding protein\n",
      "gene_26570_1 asmbl_75057 new pep maps to it without coding protein\n",
      "gene_2125 asmbl_6326 new pep maps to it without coding protein\n",
      "gene_2125 asmbl_6329 new pep maps to it without coding protein\n",
      "gene_9436 asmbl_27484 new pep maps to it without coding protein\n",
      "gene_24136 asmbl_68438 new pep maps to it which already has coding protein\n",
      "gene_2200 asmbl_6647 new pep maps to it without coding protein\n",
      "gene_2200 asmbl_6650 new pep maps to it without coding protein\n",
      "gene_1097 asmbl_3713 new pep maps to it without coding protein\n",
      "gene_12332 asmbl_35387 new pep maps to it without coding protein\n",
      "gene_14240 asmbl_40822 new pep maps to it which already has coding protein\n",
      "gene_14240 asmbl_40823 new pep maps to it which already has coding protein\n",
      "gene_10987 asmbl_32007 new pep maps to it which already has coding protein\n",
      "gene_10987 asmbl_32008 new pep maps to it which already has coding protein\n",
      "gene_10987 asmbl_32009 new pep maps to it which already has coding protein\n",
      "gene_10987 asmbl_32011 new pep maps to it which already has coding protein\n",
      "gene_16568 asmbl_46699 new pep maps to it without coding protein\n",
      "gene_16568 asmbl_46700 new pep maps to it without coding protein\n",
      "gene_27950 asmbl_78963 new pep maps to it without coding protein\n",
      "gene_27950 asmbl_78964 new pep maps to it without coding protein\n",
      "gene_6546 asmbl_19498 new pep maps to it without coding protein\n",
      "gene_28529 asmbl_80518 new pep maps to it without coding protein\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_23679 asmbl_67112 new pep maps to it without coding protein\n",
      "gene_23679 asmbl_67113 new pep maps to it without coding protein\n",
      "gene_11884 asmbl_33992 new pep maps to it without coding protein\n",
      "gene_11884 asmbl_33993 new pep maps to it without coding protein\n",
      "gene_21542 asmbl_60477 new pep maps to it which already has coding protein\n",
      "gene_23802 asmbl_67366 new pep maps to it without coding protein\n",
      "gene_23802 asmbl_67369 new pep maps to it without coding protein\n",
      "gene_23802 asmbl_67371 new pep maps to it without coding protein\n",
      "gene_23802 asmbl_67372 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21010 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21009 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21011 new pep maps to it without coding protein\n",
      "gene_11002_3 asmbl_32066 new pep maps to it without coding protein\n",
      "gene_11002_3 asmbl_32067 new pep maps to it without coding protein\n",
      "gene_11002_3 asmbl_32068 new pep maps to it without coding protein\n",
      "gene_9399 asmbl_27389 new pep maps to it without coding protein\n",
      "gene_11615 asmbl_33284 new pep maps to it without coding protein\n",
      "gene_3644 asmbl_10667 new pep maps to it without coding protein\n",
      "gene_3644 asmbl_10668 new pep maps to it without coding protein\n",
      "gene_3644 asmbl_10670 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66255 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66256 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66257 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66258 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66259 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66260 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66261 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66262 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66263 new pep maps to it without coding protein\n",
      "gene_26940 asmbl_76148 new pep maps to it without coding protein\n",
      "gene_23258_3 asmbl_65850 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65837 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65842 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65843 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65844 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65845 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65846 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65847 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65848 new pep maps to it which already has coding protein\n",
      "gene_23258_1 asmbl_65854 new pep maps to it which already has coding protein\n",
      "gene_18714_3 asmbl_52746 new pep maps to it without coding protein\n",
      "gene_18714_3 asmbl_52732 new pep maps to it without coding protein\n",
      "gene_3736 asmbl_10943 new pep maps to it which already has coding protein\n",
      "gene_28757 asmbl_81468 new pep maps to it which already has coding protein\n",
      "gene_28757 asmbl_81469 new pep maps to it which already has coding protein\n",
      "gene_26240 asmbl_74023 new pep maps to it without coding protein\n",
      "gene_2948 asmbl_8690 new pep maps to it without coding protein\n",
      "gene_2948 asmbl_8691 new pep maps to it without coding protein\n",
      "gene_2948 asmbl_8692 new pep maps to it without coding protein\n",
      "gene_29895 asmbl_84694 new pep maps to it which already has coding protein\n",
      "gene_7045 asmbl_21010 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21009 new pep maps to it without coding protein\n",
      "gene_7045 asmbl_21011 new pep maps to it without coding protein\n",
      "gene_7510 asmbl_22087 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66255 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66256 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66257 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66258 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66259 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66260 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66261 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66262 new pep maps to it without coding protein\n",
      "gene_23348 asmbl_66263 new pep maps to it without coding protein\n",
      "VQSLVDTILEDPDRVPPIDVL peptide already in known sequence asmbl_30010.p1\n",
      "VQSLVDTILEDPDRVPPIDVL peptide already in known sequence asmbl_30011.p1\n",
      "VQSLVDTILEDPDRVPPIDVL peptide already in known sequence asmbl_30012.p1\n",
      "gene_10270 asmbl_30010 new pep maps to it without coding protein\n",
      "gene_10270 asmbl_30011 new pep maps to it without coding protein\n",
      "gene_10270 asmbl_30012 new pep maps to it without coding protein\n",
      "gene_885_1 asmbl_2945 new pep maps to it without coding protein\n",
      "gene_885_1 asmbl_2946 new pep maps to it without coding protein\n",
      "gene_885_1 asmbl_2947 new pep maps to it without coding protein\n",
      "gene_885_1 asmbl_2948 new pep maps to it without coding protein\n",
      "RDEHQDFMDEQR peptide already in known sequence asmbl_45816.p1\n",
      "gene_16229 asmbl_45816 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68585 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68586 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68587 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68588 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68589 new pep maps to it without coding protein\n",
      "gene_24193 asmbl_68590 new pep maps to it without coding protein\n",
      "gene_3644 asmbl_10668 new pep maps to it without coding protein\n",
      "gene_3644 asmbl_10669 new pep maps to it without coding protein\n",
      "gene_3881 asmbl_11286 new pep maps to it without coding protein\n",
      "gene_3881 asmbl_11287 new pep maps to it without coding protein\n",
      "gene_29672 asmbl_84121 new pep maps to it without coding protein\n",
      "gene_3881 asmbl_11286 new pep maps to it without coding protein\n",
      "gene_3881 asmbl_11287 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61079 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61080 new pep maps to it without coding protein\n",
      "gene_21686 asmbl_61081 new pep maps to it without coding protein\n",
      "gene_6616_1 asmbl_19762 new pep maps to it without coding protein\n",
      "gene_6616_2 asmbl_19763 new pep maps to it without coding protein\n",
      "gene_9080 asmbl_26609 new pep maps to it without coding protein\n",
      "gene_9080 asmbl_26610 new pep maps to it without coding protein\n",
      "gene_9080 asmbl_26613 new pep maps to it without coding protein\n",
      "gene_9080 asmbl_26614 new pep maps to it without coding protein\n",
      "gene_1118 asmbl_3762 new pep maps to it without coding protein\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50485.p1\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50486.p1\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50487.p1\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50488.p1\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50489.p1\n",
      "TDYNASVSVPDSSGPER peptide already in known sequence asmbl_50490.p1\n",
      "gene_17873 asmbl_50485 new pep maps to it without coding protein\n",
      "gene_17873 asmbl_50486 new pep maps to it without coding protein\n",
      "gene_17873 asmbl_50487 new pep maps to it without coding protein\n",
      "gene_17873 asmbl_50488 new pep maps to it without coding protein\n",
      "gene_17873 asmbl_50489 new pep maps to it without coding protein\n",
      "gene_17873 asmbl_50490 new pep maps to it without coding protein\n",
      "gene_17913 asmbl_50576 new pep maps to it without coding protein\n",
      "gene_10692 asmbl_31305 new pep maps to it which already has coding protein\n",
      "gene_14290 asmbl_40932 new pep maps to it without coding protein\n",
      "gene_14290 asmbl_40933 new pep maps to it without coding protein\n",
      "gene_14290 asmbl_40934 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52487 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52488 new pep maps to it without coding protein\n",
      "gene_18635 asmbl_52486 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24866 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24867 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24868 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24869 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24874 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24870 new pep maps to it without coding protein\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_8430 asmbl_24871 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24875 new pep maps to it without coding protein\n",
      "gene_8430 asmbl_24872 new pep maps to it without coding protein\n",
      "gene_24146 asmbl_68477 new pep maps to it without coding protein\n",
      "gene_13453 asmbl_38708 new pep maps to it without coding protein\n",
      "gene_22404 asmbl_63225 new pep maps to it without coding protein\n",
      "gene_18714_3 asmbl_52746 new pep maps to it without coding protein\n",
      "gene_18714_3 asmbl_52732 new pep maps to it without coding protein\n",
      "gene_28529 asmbl_80518 new pep maps to it without coding protein\n"
     ]
    }
   ],
   "source": [
    "update_gff = pasa_path+'/05_updated.gff3'\n",
    "if not os.path.exists(update_gff):\n",
    "    # 1. generate the update assemblies\n",
    "    new_event_gff_dic = {} # {geneid:[[geneline],{assm:[lines]}]}\n",
    "    rm_genes = [] # stores which genes are modified using proteomics evidence.\n",
    "    update_assms = []\n",
    "    with open(uniq_event) as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'): continue\n",
    "            item = line.strip().split('\\t')\n",
    "            if item[0] == 'transcript gene(non CDS)':\n",
    "                new_event_gff_dic,rm_genes,update_assm = new_transcript_CDS(new_event_gff_dic,item,rm_genes,\n",
    "                                    pasa_assm_dic,pasa_pos_dic,gene_rna_dic,rna_all_dic,\n",
    "                                   gene_pos,trans_pr_pos_dic,gff_pr_dic,tr_pr_dic)\n",
    "                update_assms += update_assm\n",
    "    # 2. merge updated assemblies with the old one\n",
    "    # update all_gff 1. copy the original gff. 2. remove the genes that should be updated. 3. add updated genes.\n",
    "    cp_all_gff = all_gff.copy()\n",
    "    for rm_gene in set(rm_genes):\n",
    "        del cp_all_gff[rm_gene]\n",
    "    cp_all_gff.update(new_event_gff_dic)\n",
    "    # output the updated gff file\n",
    "    with open(update_gff,'w') as f:\n",
    "        for gene,v in cp_all_gff.iteritems():\n",
    "            v[0][3] = str(v[0][3])\n",
    "            v[0][4] = str(v[0][4])\n",
    "            f.write('\\t'.join(v[0])+'\\n')\n",
    "            for assm,items in v[1].iteritems():\n",
    "                for item in items:\n",
    "                    item[3] = str(item[3])\n",
    "                    item[4] = str(item[4])\n",
    "                    f.write('\\t'.join(item)+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "136 genes are updated. 208 transcripts map without coding proteins. 38 transcripts already have coding protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# add frame information\n",
    "def add_cds_frame(in_gff,frame_gff):\n",
    "    with open(frame_gff,'w') as out, open(in_gff) as inf:\n",
    "        pre_pr = ''\n",
    "        for line in inf:\n",
    "            if line.startswith('#'):\n",
    "                out.write(line)\n",
    "            else:\n",
    "                item = line.split('\\t')\n",
    "                if item[2] == 'CDS':\n",
    "                    pr = re.search('(?<=protein_id=).+?(?=$|;)',item[8]).group(0)\n",
    "                    if pr != pre_pr:\n",
    "                        pre_pr = pr\n",
    "                        item[7] = '0'\n",
    "                        n = abs(int(item[3])-int(item[4]))+1\n",
    "                    else:\n",
    "                        item[7] = str(n % 3)\n",
    "                        n += abs(int(item[3])-int(item[4]))+1\n",
    "                    out.write('\\t'.join(item))\n",
    "                else:\n",
    "                    out.write(line)\n",
    "\n",
    "update_gff = pasa_path+'/05_updated.gff3'\n",
    "up_cds_frame_gff = pasa_path+'/06_updated_cds_frame.gff3'\n",
    "if not os.path.exists(up_cds_frame_gff):\n",
    "    add_cds_frame(update_gff,up_cds_frame_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# extract proteins from gff\n",
    "import sarge\n",
    "up_cds_frame_pr = pasa_path+'/06_updated_cds_frame_pr.fa'\n",
    "if not os.path.exists(up_cds_frame_pr):\n",
    "    cmd = ('gffread {gff} -g {ref} -y {fa}').format(gff=up_cds_frame_gff,\n",
    "                                                    ref=ref_fa,\n",
    "                                                    fa=up_cds_frame_pr)\n",
    "    sarge.run(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#####  Functional annotation of the new added proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#--------- 1. extract low quality proteins (new added) based on the transcripts id mapping.\n",
    "low_qual_pr_fn = pasa_path+'/07_update_low_qual_pr.fa'\n",
    "if not os.path.exists(low_qual_pr_fn):\n",
    "    update_pr_index = SeqIO.index(up_cds_frame_pr,'fasta')\n",
    "    with open(low_qual_pr_fn,'w') as f:\n",
    "        for assm in update_pr_index:\n",
    "            if assm in update_assms:\n",
    "                SeqIO.write(update_pr_index[assm],f,'fasta')\n",
    "#---------- 2. blastp these proteins to uniprot\n",
    "uniprot_db = pasa_path+'/uniprot_blastDb/uniprot'\n",
    "low_qual_pr2uni = pasa_path + '/08_update_low_qual_pr2uni.txt'\n",
    "if not os.path.exists(low_qual_pr2uni):\n",
    "    cmd = ('blastp -db {uni} -query {q} -out {out} -outfmt 6 -num_alignments 1 -num_threads 9 -evalue 1').format(\n",
    "            uni=uniprot_db,q=low_qual_pr_fn,out=low_qual_pr2uni)\n",
    "    sarge.run(cmd)\n",
    "#---------- 3. add gene name to the blastp mapping results.\n",
    "# first build {uniprot_accession: gene name } dictionary\n",
    "uni_id_map_fn = pasa_path + '/uniprot_idmap.txt'\n",
    "df = pd.read_csv(uni_id_map_fn,sep='\\t',header=None,names=['acc','name'])\n",
    "df['name'] = df['name'].map(lambda x: str(x).upper())\n",
    "acc_name_dic = df.set_index('acc')['name'].to_dict()\n",
    "# second build {transcripts:gene name} dictionary\n",
    "df = pd.read_csv(low_qual_pr2uni,sep='\\t',header=None)\n",
    "df['acc'] = df[1].map(lambda x: x.split('|')[1])\n",
    "df['gene'] =df['acc'].map(lambda x:acc_name_dic[x] if x in acc_name_dic else x)\n",
    "low_assm_gnm_dic = df.set_index(0)['gene'].to_dict()\n",
    "low_assm_acc_dic = df.set_index(0)[1].to_dict()\n",
    "# third output the new gene name\n",
    "update_gnm_fn = pasa_path + '/10_update_gnm.txt'\n",
    "if not os.path.exists(update_gnm_fn):\n",
    "    with open(update_gnm_fn,'w') as out:\n",
    "        for k,v in low_assm_gnm_dic.iteritems():\n",
    "            out.write(k+'\\t'+v+'\\t'+low_assm_acc_dic[k]+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# update gene name in the updated gff file\n",
    "# import pdb;pdb.set_trace()\n",
    "final_update_gff = pasa_path+'/09_updated_final.gff3'\n",
    "up_cds_frame_gff = pasa_path+'/06_updated_cds_frame.gff3'\n",
    "if not os.path.exists(final_update_gff):\n",
    "    with open(up_cds_frame_gff) as in_f, open(final_update_gff,'w') as out:\n",
    "        for line in in_f:\n",
    "            if line.startswith('#'):\n",
    "                out.write(line)\n",
    "                continue\n",
    "            item = line.strip().split('\\t')\n",
    "            if item[2] == 'gene':\n",
    "                gene = line\n",
    "                n = 1\n",
    "            elif item[2] in ['mRNA','lncRNA']:\n",
    "                rnaid = re.search('(?<=ID=).+?(?=;|$)',line).group(0)\n",
    "                if rnaid in low_assm_gnm_dic:\n",
    "                    if n == 1:\n",
    "                        gene = re.sub('(?<=Name=).+?(?=;|$)',low_assm_gnm_dic[rnaid],gene)\n",
    "                        out.write(gene)\n",
    "                    line = re.sub('(?<=Name=).+?(?=;|$)',low_assm_gnm_dic[rnaid],line)\n",
    "                else:\n",
    "                    if n == 1:\n",
    "                        out.write(gene)\n",
    "                out.write(line)\n",
    "                n += 1\n",
    "            else:\n",
    "                rnaid = re.search('(?<=Parent=).+?(?=;|$)',line).group(0)\n",
    "                if rnaid in low_assm_gnm_dic:\n",
    "                    line = re.sub('(?<=Name=).+?(?=;|$)',low_assm_gnm_dic[rnaid],line)\n",
    "                out.write(line)\n",
    "                n += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "update_final_pr = pasa_path + '/09_updated_final_pr.fa'\n",
    "if not os.path.exists(update_final_pr):\n",
    "    cmd = ('gffread {gff} -g {ref} -y {fa}').format(gff=final_update_gff,\n",
    "                                                    ref=ref_fa,\n",
    "                                                    fa=up_cds_frame_pr)\n",
    "    sarge.run(cmd)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Hide code",
  "hide_code_all_hidden": false,
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
